benchmarks/bench_isolation_forest_predict.py

```diff
@@ -0,0 +1,186 @@
+    print(merged_data["n_jobs"].unique())
+    ax = axes[0]
+    sns.lineplot(
+        data=merged_data,
+        x="n_samples_test",
+        y="predict_time_pr",
+        hue="n_jobs",
+        style="n_jobs",
+        markers="o",
+        # dashes=False,
+        ax=ax,
+        legend="full",
+    )
+    ax.set_title(f"Predict Time vs. n_samples_test - {pr_name} branch")
+    ax.set_ylabel("Predict Time (Seconds)")
+    ax.set_xlabel("n_samples_test")
+
+    ax = axes[1]
+    sns.lineplot(
+        data=merged_data,
+        x="n_samples_test",
+        y="predict_time_main",
+        hue="n_jobs",
+        style="n_jobs",
+        markers="X",
+        dashes=True,
+        ax=ax,
+        legend=None,
+    )
+    ax.set_title(f"Predict Time vs. n_samples_test - {main_name} branch")
+    ax.set_ylabel("Predict Time")
+    ax.set_xlabel("n_samples_test")
+
+    # Adjust layout and display the plots
+    plt.tight_layout()
+    fig.savefig(image_path, bbox_inches="tight")
+    print(f"Saved image to {image_path}")
+
+
+def main():
+    random_state = 1
+
+    results = defaultdict(list)
+
+    # Loop over all datasets for fitting and scoring the estimator:
+    n_samples_train = 1000
+    for n_samples_test in [
+        1000,
+        10000,
+        50000,
+    ]:
+        for n_features in [10, 100, 1000]:
+            for contamination in [0.01, 0.1, 0.5]:
+                for n_jobs in [1, 2, 3, 4]:
+                    X_train, X_test = get_data(
+                        n_samples_train,
+                        n_samples_test,
+                        n_features,
+                        contamination,
+                        random_state,
+                    )
+
+                    print("--- Fitting the IsolationForest estimator...")
+                    model = IsolationForest(n_jobs=-1, random_state=random_state)
+                    tstart = time()
+                    model.fit(X_train)
+                    fit_time = time() - tstart
+
+                    tstart = time()
+
+                    # clearcache
+                    for _ in range(1000):
+                        1 + 1
+                    with parallel_config("threading", n_jobs=n_jobs):
+                        tstart = time()
+                        model.decision_function(X_test)  # the lower, the more abnormal
+                        predict_time = time() - tstart
+
+                    results["predict_time"].append(predict_time)
+                    results["fit_time"].append(fit_time)
+                    results["n_samples_train"].append(n_samples_train)
+                    results["n_samples_test"].append(n_samples_test)
+                    results["n_features"].append(n_features)
+                    results["contamination"].append(contamination)
+                    results["n_jobs"].append(n_jobs)
+
+    # df = pd.DataFrame(results)
+    # df.to_csv("~/bench_results_forest/pr-threading.csv", index=False)
```

Comment:
This is commended out. Can this benchmark function have two commands? One for running the benchmark and saving the results to disk and another one for plotting the results?

Reply:
I fixed this and made two commands for the benchmark.